helper-cli/src/main/kotlin/utils/PathExcludeGenerator.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) 2022 EPAM Systems, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.helper.utils
+
+import java.io.File
+
+import org.ossreviewtoolkit.model.config.PathExclude
+import org.ossreviewtoolkit.model.config.PathExcludeReason.BUILD_TOOL_OF
+import org.ossreviewtoolkit.model.config.PathExcludeReason.DOCUMENTATION_OF
+import org.ossreviewtoolkit.model.config.PathExcludeReason.TEST_OF
+
+/**
+ * This class generates path excludes based on the set file paths present in the source tree.
+ */
+internal object PathExcludeGenerator {
+    /**
+     * Return path excludes which likely but not necessarily apply to a source tree containing all given [filePaths]
+     * which must be relative to the root directory of the source tree.
+     */
+    fun generatePathExcludes(filePaths: Set<String>): List<PathExclude> {
+        val files = filePaths.mapTo(mutableSetOf()) { File(it) }
+        val dirs = getAllDirs(files)
+
+        val pathExcludes = mutableSetOf<PathExclude>()
+
+        dirs.forEach { dir ->
+            PATH_EXCLUDES_REASON_FOR_DIR_NAME[dir.name]?.let { reason ->
+                pathExcludes += PathExclude(
+                    pattern = "${dir.path}/**",
+                    reason = reason
+                )
+            }
+        }
+
+        val filesForPathExcludes = pathExcludes.associateWith { pathExcludeExclude ->
+            files.filter { pathExcludeExclude.matches(it.path) }.toSet()
+        }
+
+        return greedySetCover(filesForPathExcludes).toList()
+    }
+}
+
+private fun getAllDirs(files: Set<File>): Set<File> {
+    val result = mutableSetOf<File>()
+
+    files.forEach { file ->
+        var dir = file.parentFile
+
+        while (dir != null) {
+            result += dir
+            dir = dir.parentFile
+        }
+    }
+
+    return result
+}
+
+private val PATH_EXCLUDES_REASON_FOR_DIR_NAME = mapOf(
```

Comment:
Could we make below mapping a config file in YAML?

Reply:
Yes. Would you be fine doing this in a separate PR later on?

Reply:
@tsteenbe let's sort this out later on, I'll experiment with the more complex heuristic first and try out having a fun test. Then we can still extract the patterns to a config file.